sched/event/event_wait.c

```diff
@@ -0,0 +1,211 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/event.h>
+#include <nuttx/sched.h>
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxevent_wait_internal
+ *
+ * Description:
+ *   Wait for all of the specified events for the specified tick time.
+ *
+ *   This routine waits on event object event until all of the specified
+ *   events have been delivered to the event object, or the maximum wait time
+ *   timeout has expired. A thread may wait on up to 32 distinctly numbered
+ *   events that are expressed as bits in a single 32-bit word.
+ *
+ * Input Parameters:
+ *   event  - Address of the event object
+ *   events - Set of events to wait
+ *          - Set events to 0 will indicate wait from any events
+ *   delay  - Ticks to wait from the start time until the event is
+ *            posted.  If ticks is zero, then this function is equivalent
+ *            to nxevent_trywait().
+ *
+ * Returned Value:
+ *   This is an internal OS interface and should not be used by applications.
+ *   Return of matching events upon success.
+ *   0 if matching events were not received within the specified time.
+ *
+ ****************************************************************************/
+
+static nxevent_mask_t nxevent_wait_internal(FAR nxevent_t *event,
+                                            nxevent_mask_t events,
+                                            uint32_t delay)
+{
+  FAR struct tcb_s *rtcb;
+  irqstate_t flags;
+  int ret;
+
+  DEBUGASSERT(event != NULL && up_interrupt_context() == false);
+
+  flags = enter_critical_section();
+
+  if ((events == 0 && event->events) ||
+      (events != 0 && event->events == events))
+    {
+      events = event->events;
+      event->events = 0;
+    }
+  else
+    {
+      rtcb         = nxsched_self();
+      rtcb->events = events;
+
+      /* Remove pending counter */
+
+      nxsem_reset(&event->sem, 0);
+
+      if (delay == UINT32_MAX)
+        {
+          ret = nxsem_wait_uninterruptible(&event->sem);
+        }
+      else if (delay == 0)
+        {
+          ret = nxsem_trywait(&event->sem);
+        }
+      else
+        {
+          ret = nxsem_tickwait_uninterruptible(&event->sem, delay);
+        }
+
+      if (ret == 0)
+        {
+          if (events != 0)
+            {
+              events &= event->events;
+            }
+          else
+            {
+              events = event->events;
```

Comment:
why clear all event?

Reply:
all events are cleared on each time of wait returns, which ensures that the notifier call nxevent_post() could set same event mask again to avoid missing this event

Reply:
we should use the same condition from line 69-70, so where loop may better.

Reply:
> all events are cleared on each time of wait returns, which ensures that the notifier call nxevent_post() could set same event mask again to avoid missing this event

Does this mean that events the occur between the time that the wait returns and the mask is reset are lost?  Is the the desired behavior?  Couldn't that result in those events being lost?  Should they pend instead?

Reply:
@patacongo thanks for your review, you are right, there may be potential race conditions If all events are cleared in multi-tasking case, I will update my implementation to keep the pending event, please review again later.